src/devices/Ad7193/Ad7193.cs

```diff
@@ -0,0 +1,855 @@
+
+            return raw;
+        }
+
+        /// <summary>
+        /// Reads a single value on the selected channel
+        /// </summary>
+        /// <param name="channel">Channel index</param>
+        /// <returns></returns>
+        public uint? ReadSingleADCValue(Channel channel)
+        {
+            ActiveChannels = channel;
+
+            StartSingleConversion();
+
+            WaitForADC();
+
+            return ReadADCValue();
+        }
+
+        /// <summary>
+        /// Converts the ADC result to Volts
+        /// </summary>
+        /// <param name="adcValue">The raw ADC result</param>
+        /// <returns></returns>
+        private double RawValueToVoltage(uint adcValue)
+        {
+            int pgaGain = 1;
+            switch (PGAGain)
+            {
+                case Gain.X1:
+                    pgaGain = 1;
+                    break;
+                case Gain.X8:
+                    pgaGain = 8;
+                    break;
+                case Gain.X16:
+                    pgaGain = 16;
+                    break;
+                case Gain.X32:
+                    pgaGain = 32;
+                    break;
+                case Gain.X64:
+                    pgaGain = 64;
+                    break;
+                case Gain.X128:
+                    pgaGain = 128;
+                    break;
+            }
+
+            double voltage = 0;
+
+            // 0 - bipolar (ranges from ±19.53 mV to ±2.5 V) ; 1 - unipolar (ranges from 0 mV to 19.53 mV to 0 V to 2.5 V)
+            if (Unipolar)
+            {
+                voltage = (double)adcValue / 16777216.0;
+            }
+            else if (!Unipolar)
+            {
+                voltage = ((double)adcValue / 8388608.0) - 1.0;
+            }
+
+            voltage *= (VReference / pgaGain);
+
+            return (voltage);
+        }
+
+        /// <summary>
+        /// Converts the ADC result to Celsius
+        /// </summary>
+        /// <param name="adcValue">The raw ADC result</param>
+        /// <returns></returns>
+        protected Temperature ADCValueToTemperature(ulong adcValue)
+        {
+            double degreeCelsius = ((float)(adcValue - 0x0080_0000) / 2815.0f) - 273.0f;
+            return Temperature.FromCelsius(degreeCelsius);
+        }
+
+        /// <summary>
+        /// Reads the value of a register
+        /// </summary>
+        protected uint GetRegisterValue(Register register)
+        {
+            byte registerAddress = (byte)register;
+            byte byteNumber = _registerSize[registerAddress];
+            byte[] writeBuffer = new byte[byteNumber + 1];
+
+            byte commandByte = (byte)((byte)CommunicationsRegisterBits.ReadOperation | GetCommAddress(registerAddress));
+            writeBuffer[0] = commandByte;
+
+            byte[] readBuffer = new byte[writeBuffer.Length];
+            lock (_spiTransferLock)
```

Comment:
usually we make classes not thread safe and the caller can lock them if they want. I haven't seen any thread being spawned in the ctor - might make sense to remove the locks overall since that will match what we do elsewhere

Reply:
I think I had some trouble with threads while reading the value when I used the event handler in the test application, and that's why I decided to use the lock in the driver. 
It would work if I used it in the sample application as well, but then everyone who uses the event-based value acquisition should know that they need to use locks.
This is a safer solution, although I understand it is not critical, since everyone could do the lock manually when they use this driver.
Would you still prefer to remove it from here and let the users do the locking?

Reply:
I'd prefer we removed all the threads and helpers which are not essential. This logic should be made on top of ADC. Even though we currently don't have abstraction we could plug that into it's still better to design APIs so that they don't have anything which adds too much complexity